src/generic.jl

```diff
@@ -120,6 +120,37 @@ function _pairwise!(r::AbstractMatrix, metric::SemiMetric, a::AbstractMatrix)
     r
 end
 
+function _pairwise!(r::AbstractMatrix, metric::PreMetric,
+                    a::AbstractVector, b::AbstractVector=a)
+    require_one_based_indexing(r)
+    na = length(a)
+    nb = length(b)
+    size(r) == (na, nb) || throw(DimensionMismatch("Incorrect size of r."))
+    @inbounds for (j, bj) = enumerate(b)
+        for (i, ai) = enumerate(a)
+            r[i, j] = metric(ai, bj)
+        end
+    end
+    r
+end
+
+function _pairwise!(r::AbstractMatrix, metric::SemiMetric, a::AbstractVector)
+    require_one_based_indexing(a, r)
+    n = length(a)
+    size(r) == (n, n) || throw(DimensionMismatch("Incorrect size of r."))
+    @inbounds for j = 1:n
+        for i = 1:(j - 1)
+            r[i, j] = metric(a[i], a[j])
+        end
+        r[j, j] = 0
+        for i = (j + 1):n
+            r[i, j] = r[j, i]   # leveraging the symmetry of SemiMetric
+        end
+    end
```

Comment:
I think it should be
```suggestion
        for i = 1:(j - 1)
            r[i, j] = r[j, i]   # leveraging the symmetry of SemiMetric
        end
        r[j, j] = 0
        for i = (j + 1):n
            r[i, j] = metric(a[i], a[j])
        end
    end
```
otherwise you copy from places that you haven't filled yet, if I'm not mistaken.

Reply:
Oh my bad, unfortunate copy/paste

Reply:
Why not use this order of the loops? For each column index `j`, you would start from row `i in 1:j-1`, then fill `r[j,j]`, and then contiguously continue with `i in j+1:n`. I meant to switch the order of the loops, viewed as entire blocks, including their loop body.

Reply:
Ok I think I got confused at first but it's more clear now. I will change it.